cpp/src/arrow/array.h

```diff
@@ -239,6 +239,11 @@ class ARROW_EXPORT Array {
   /// \return PrettyPrint representation of array suitable for debugging
   std::string ToString() const;
 
+  Status cast(const std::shared_ptr<DataType>& type, std::shared_ptr<Array>* out);
+
+  template <typename T>
+  Status castTyped(std::shared_ptr<Array>* out);
```

Comment:
Casting is a can of worms, I don't think it makes sense to add these methods here yet:

* Requires memory allocations
* Two-way dynamic dispatch (input type x output type)

The `MakeDictionaryArray` function below will suffice for the scope of this issue